### ISSUE TITLE
Set of tuples with string crashes

### ISSUE THREAD

**Issue by ehsantn** [numba/numba#4577]
## Reporting a bug

- [x] I am using the latest ~~released~~ master version of Numba (most recent is visible in
 the change log (https://github.com/numba/numba/blob/master/CHANGE_LOG).
- [x] I have included below a minimal working reproducer (if you are unsure how
 to write one see http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports).

Using tuples containing strings as keys in Set containers can cause crashes or invalid output. Example code below:
```python
@numba.njit
def f(A):
    s = set()
    for i in range(len(A)):
        s.add((str(A[i]), str(A[i])))
    print(s)

A = np.array(['abc', 'aa', 'ccd', 'ggvf', 'aa', 'abc'])
f(A)
```

**Comment by ehsantn** [comment]
Adding an `incref` call [here](https://github.com/numba/numba/blob/master/numba/targets/setobj.py#L433) seems to fix it for this particular example, so I assume this is a refcount issue. Looks like refcounting is not handled anywhere in the implementation.

**Comment by stuartarchibald** [comment]
Thanks for the report. Likely cause is the `set` impl was written before the time of having anything hashable and refcounted.

**Comment by stuartarchibald** [comment]
@ehsantn do you have an example of a dictionary segfaulting with similar? Thanks.

**Comment by ehsantn** [comment]
@stuartarchibald No, looks like Dict is fine. My code was using both, and somehow looked like Dict has problems too. However, everything seems to work if I add incref/decref to Set code.

**Comment by stuartarchibald** [comment]
@ehsantn ah ok, thanks for confirming. Guess it's just the noted issue then.

**Comment by stuartarchibald** [comment]
Note: Mainline as of 0.55 reports:
```
numba.core.errors.NumbaValueError: Failed in nopython mode pipeline (step: native lowering)
Use of reference counted items in 'set()' is unsupported, offending type is: 'UniTuple(unicode_type x 2)'.
```
for the MWR.

**Comment by stuartarchibald** [comment]
Closing, fixed by #7780.
